File: src/algorithms/leaves.ts
import {getPath} from "../metrics/tree";
import {ExtendedLeafData, LEAF_DATA_NUMERIC_FIELDS, LeafData} from '../types/data';
import {TreeNode} from '../types/node';


export const getLeavesData = (node: TreeNode): LeafData[] =>
	node.children.length
		? node.children.flatMap(getLeavesData)
		: [node.data as LeafData];

type MinMaxRet = {
	minimumsOfNumericFields: number[],
	maximumsOfNumericFields: number[],
};

export const calculateMinMaxOfNumericFields = (leavesData: LeafData[]): MinMaxRet => {
	const minimumsOfNumericFields = Array(LEAF_DATA_NUMERIC_FIELDS.length).fill(Infinity);
	const maximumsOfNumericFields = Array(LEAF_DATA_NUMERIC_FIELDS.length).fill(-Infinity);
	for (let i = 0; i < leavesData.length; ++i) {
		for (let j = 0; j < LEAF_DATA_NUMERIC_FIELDS.length; ++j) {
			const field: string = LEAF_DATA_NUMERIC_FIELDS[j];
			minimumsOfNumericFields[j] = Math.min(leavesData[i][field], minimumsOfNumericFields[j]);
			maximumsOfNumericFields[j] = Math.max(leavesData[i][field], maximumsOfNumericFields[j]);
		}
	}
	return {minimumsOfNumericFields, maximumsOfNumericFields};
};

export const normalizeLeavesData = (leavesData: LeafData[]): ExtendedLeafData[] => {
	const {minimumsOfNumericFields, maximumsOfNumericFields} = calculateMinMaxOfNumericFields(leavesData);
	const normalizedLeavesData: ExtendedLeafData[] = leavesData.map((leafData: LeafData): ExtendedLeafData => new ExtendedLeafData(leafData));
	for (let i = 0; i < leavesData.length; ++i) {
		for (let j = 0; j < LEAF_DATA_NUMERIC_FIELDS.length; ++j) {
			const field: string = LEAF_DATA_NUMERIC_FIELDS[j];
			const min: number = minimumsOfNumericFields[j];
			const max: number = maximumsOfNumericFields[j];
			normalizedLeavesData[i].normalized[field] = (leavesData[i][field] - min) / (max - min);
		}
	}
	return normalizedLeavesData;
};

export const fillPaths = (root: TreeNode, extendedLeavesData: ExtendedLeafData[]) => {
	extendedLeavesData.forEach(
		(extendedLeafData: ExtendedLeafData) => {
			extendedLeafData.path = getPath(root, extendedLeafData.source.name);
		}
	);
};

File: src/678-va-dialog/algorithms/leaves.py
from math import inf
from typing import List, Tuple

from algorithms.flat_map import flat_map
from metrics.tree import get_path
from ttypes.data import LeafData, ExtendedLeafData, LEAF_DATA_NUMERIC_FIELDS
from ttypes.node import TreeNode


def get_leaves_data(node: TreeNode) -> List[LeafData]:
    if node['children']:
        return flat_map(get_leaves_data, node['children'])
    return [node['data']]


def calculate_min_max_of_numeric_fields(leaves_data: List[LeafData]) -> Tuple[List[float], List[float]]:
    minimums_of_numeric_fields = [inf] * len(LEAF_DATA_NUMERIC_FIELDS)
    maximums_of_numeric_fields = [-inf] * len(LEAF_DATA_NUMERIC_FIELDS)
    for leaf_data in leaves_data:
        for j, field in enumerate(LEAF_DATA_NUMERIC_FIELDS):
            minimums_of_numeric_fields[j] = min(leaf_data[field], minimums_of_numeric_fields[j])
            maximums_of_numeric_fields[j] = max(leaf_data[field], maximums_of_numeric_fields[j])
    return minimums_of_numeric_fields, maximums_of_numeric_fields


def normalize_leaves_data(leaves_data: List[LeafData]) -> List[ExtendedLeafData]:
    minimums_of_numeric_fields, maximums_of_numeric_fields = calculate_min_max_of_numeric_fields(leaves_data)
    normalized_leaves_data: List[ExtendedLeafData] = [ExtendedLeafData(leaf_data) for leaf_data in leaves_data]
    for j, field in enumerate(LEAF_DATA_NUMERIC_FIELDS):
        min_j: float = minimums_of_numeric_fields[j]
        max_j: float = maximums_of_numeric_fields[j]
        for i, leaf_data in enumerate(leaves_data):
            normalized_leaves_data[i].normalized[field] = (leaf_data[field] - min_j) / (max_j - min_j)
    return normalized_leaves_data


def fill_paths(root: TreeNode, extended_leaves_data: List[ExtendedLeafData]):
    for extended_leaf_data in extended_leaves_data:
        extended_leaf_data.path = get_path(root, extended_leaf_data.source['name'])
